From 7774f8da6f85fe21b17ccbce6993579563ce852f Mon Sep 17 00:00:00 2001
From: Yu Kuai <yukuai3@huawei.com>
Date: Wed, 22 Mar 2023 11:59:26 +0800
Subject: [PATCH 394/780] [Backport] block: don't set GD_NEED_PART_SCAN if scan
 partition failed

mainline inclusion
from mainline-v6.3-rc6
commit 3723091ea1884d599cc8b8bf719d6f42e8d4d8b1
category: bugfix
bugzilla: 188884

Reference: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?h=v6.4-rc7&id=3723091ea1884d599cc8b8bf719d6f42e8d4d8b1

--------------------------------

Currently if disk_scan_partitions() failed, GD_NEED_PART_SCAN will still
set, and partition scan will be proceed again when blkdev_get_by_dev()
is called. However, this will cause a problem that re-assemble partitioned
raid device will creat partition for underlying disk.

Test procedure:

mdadm -CR /dev/md0 -l 1 -n 2 /dev/sda /dev/sdb -e 1.0
sgdisk -n 0:0:+100MiB /dev/md0
blockdev --rereadpt /dev/sda
blockdev --rereadpt /dev/sdb
mdadm -S /dev/md0
mdadm -A /dev/md0 /dev/sda /dev/sdb

Test result: underlying disk partition and raid partition can be
observed at the same time

Note that this can still happen in come corner cases that
GD_NEED_PART_SCAN can be set for underlying disk while re-assemble raid
device.

Fixes: e5cfefa97bcc ("block: fix scan partition for exclusively open device again")
Reviewed-by: Jan Kara <jack@suse.cz>
Reviewed-by: Ming Lei <ming.lei@redhat.com>
Signed-off-by: Yu Kuai <yukuai3@huawei.com>
Signed-off-by: Jens Axboe <axboe@kernel.dk>

Conflicts:
  block/genhd.c
Signed-off-by: Li Lingfeng <lilingfeng3@huawei.com>
---
 block/genhd.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/block/genhd.c b/block/genhd.c
index 3c34dd011585..83f407e1563a 100644
--- a/block/genhd.c
+++ b/block/genhd.c
@@ -770,6 +770,13 @@ int disk_scan_partitions(struct gendisk *disk, fmode_t mode)
 	else
 		blkdev_put(bdev, mode & ~FMODE_EXCL);
 
+	/*
+	 * If blkdev_get_by_dev() failed early, GD_NEED_PART_SCAN is still set,
+	 * and this will cause that re-assemble partitioned raid device will
+	 * creat partition for underlying disk.
+	 */
+	clear_bit(GD_NEED_PART_SCAN, &disk->state);
+
 	if (!(mode & FMODE_EXCL)) {
 		bd_abort_claiming(claim, claim, disk_scan_partitions);
 		bdput(claim);
-- 
2.17.1

